src/common/dav_v2/heap.c

```diff
@@ -1700,25 +1752,28 @@ heap_create_evictable_mb(struct palloc_heap *heap, uint32_t *mb_id)
 	heap->rt->zones_exhausted_e--;
 	if (heap->rt->zones_unused_first > zone_id)
 		heap->rt->zones_unused_first = zone_id;
-	return -1;
+	rc = -1;
+
+out:
+	heap->rt->mb_create_waiters--;
+	D_ASSERT((int)heap->rt->mb_create_waiters >= 0);
+	if (heap->rt->mb_create_waiters) {
+		D_ASSERT(store->stor_ops->so_waitqueue_wakeup != NULL);
+		store->stor_ops->so_waitqueue_wakeup(heap->rt->mb_create_wq, false);
+	}
+	return rc;
 }
 
 int
 heap_get_evictable_mb(struct palloc_heap *heap, uint32_t *mb_id)
 {
 	struct mbrt       *mb;
 	int                ret;
-	struct umem_store *store = heap->layout_info.store;
-
-	heap->rt->mb_create_waiters++;
-	if (heap->rt->mb_create_waiters > 1) {
-		D_ASSERT(store->stor_ops->so_waitqueue_wait != NULL);
-		store->stor_ops->so_waitqueue_wait(heap->rt->mb_create_wq, false);
-	}
```

Comment:
So the key difference of moving the serializing section from here into the heap_create_evictable_mb() is that while creating a new EMB, some space on existing EMBs could be freed, so that heap_get_evictable_mb() could move forward to use these existing EMBs?

Reply:
By moving the serialization code here I could notice a reduction in number of waiters.  Once an Active Evictable MB is set, subsequent ULTs will return immediately. Previously new ULTs will wait till previous waiters are woken up.

Reply:
oh, right. I got it.